### PR TITLE
sybsys: nfc: NFC T4T APDU

### PR DESCRIPTION
**Pull request by KAGA164**
Please review only last three PR.

### CODE REVIEW

**Comment by KAGA164** [comment]
@grochu @kapi-no Please review

File: subsys/nfc/t4t/apdu.c [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+#define STATUS_SIZE           2U
+
+#define MSB_16(a) (((a) & 0xFF00) >> 8)
+#define LSB_16(a) ((a) & 0x00FF)
+
+static u16_t nfc_t4t_apdu_comm_size_calc(const struct nfc_t4t_apdu_comm *cmd_apdu)
+{
+	u16_t res = CLASS_TYPE_SIZE + INSTRUCTION_TYPE_SIZE + PARAMETER_SIZE;
+
+	if (cmd_apdu->data.buff != NULL) {
```
Comment (kapi-no) [inline subsys/nfc/t4t/apdu.c]: In Zephyr CS, it would be:
```if (cmd_apdu->data.buff)```

File: subsys/nfc/t4t/apdu.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		}
+	}
+
+	return res;
+}
+
+static int nfc_t4t_apdu_comm_args_validate(const struct nfc_t4t_apdu_comm *cmd_apdu,
+					   u8_t *raw_data, u16_t *len)
+{
+	if ((cmd_apdu == NULL) || (raw_data == NULL) || (len == NULL)) {
```
Comment (kapi-no) [inline subsys/nfc/t4t/apdu.c]: ```!cma_apdu```
and so on

File: subsys/nfc/t4t/apdu.c [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	}
+
+	return 0;
+}
+
+static int nfc_t4t_apdu_resp_args_validate(const struct nfc_t4t_apdu_resp *resp_apdu,
+					   const u8_t *raw_data,
+					   u16_t len)
+{
+	if ((resp_apdu == NULL) || (raw_data == NULL))	{
```
Comment (kapi-no) [inline subsys/nfc/t4t/apdu.c]: ditto

File: subsys/nfc/t4t/apdu.c [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+	return 0;
+}
+
+
+void nfc_t4t_apdu_resp_printout(struct nfc_t4t_apdu_resp *resp_apdu)
+{
+	LOG_INF("R-APDU status: %4X ", resp_apdu->status);
+
+	if (resp_apdu->data.buff != NULL) {
```
Comment (kapi-no) [inline subsys/nfc/t4t/apdu.c]: ditto

File: include/nfc/t4t/apdu.rst [hunk 5]
```diff
@@ -0,0 +1,62 @@
+.. _nfc_t4t_apdu:
+
+NFC T4T APDU READER/WRITER
+##########################
+
+Special communication units, called application protocol data units (APDUs), are used when exchanging data with a Type 4 Tag platform. Command APDUs (C-APDUs) are the commands sent from the NFC reader, while the response APDUs (R-APDUs) are the responses to a specific command received by NFC reader from the tag.
+
+The APDU reader/writer module provides functions to encode a C-APDU according to its descriptor and to decode a raw R-APDU data into the appropriate descriptor.
+
+Can be used with :ref:`st25r3911b_nfc_readme` to data exchange with NFC Type 4 Tag.
```
Comment (kapi-no) [inline include/nfc/t4t/apdu.rst]: My suggestion
```suggestion
Can be used with :ref:`st25r3911b_nfc_readme` to exchange NFC Type 4 Tag data.
```

Comment (KAGA164) [reply include/nfc/t4t/apdu.rst]: Ok. Changed

File: subsys/nfc/t4t/apdu.c (hunk 4 above)
Comment (KAGA164) [reply subsys/nfc/t4t/apdu.c]: Fixed

File: subsys/nfc/t4t/apdu.c (hunk 3 above)
Comment (KAGA164) [reply subsys/nfc/t4t/apdu.c]: Fixed

File: subsys/nfc/t4t/apdu.c (hunk 2 above)
Comment (KAGA164) [reply subsys/nfc/t4t/apdu.c]: Fixed

File: subsys/nfc/t4t/apdu.c (hunk 1 above)
Comment (KAGA164) [reply subsys/nfc/t4t/apdu.c]: Fixed

**Comment by KAGA164** [comment]
@b-gent please review documentation

**Comment by KAGA164** [comment]
@pdunaj please review

File: subsys/nfc/t4t/apdu.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	if (comm_apdu_len > *len) {
+		return -ENOMEM;
+	}
+
+	*len = comm_apdu_len;
+
+	/* Start to encode described CAPDU in the buffer. */
+	*raw_data++ = cmd_apdu->class_byte;
+	*raw_data++ = cmd_apdu->instruction;
+	*raw_data++ = MSB_16(cmd_apdu->parameter);
```
Comment (pdunaj) [inline subsys/nfc/t4t/apdu.c]: please use `sys_put_be16` instead of custom macros

File: subsys/nfc/t4t/apdu.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	/* Validate passed arguments. */
+	err = nfc_t4t_apdu_resp_args_validate(resp_apdu, raw_data, len);
+	if (err) {
+		return err;
+	}
+
+	nfc_t4t_apdu_resp_clear(resp_apdu);
+
+	/*  Optional data field is present in RAPDU. */
+	if (len != STATUS_SIZE) {
```
Comment (pdunaj) [inline subsys/nfc/t4t/apdu.c]: `len > STATUS_SIZE`?

File: subsys/nfc/t4t/apdu.c [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+		resp_apdu->data.len = len - STATUS_SIZE;
+		resp_apdu->data.buff = (u8_t *)raw_data;
```
Comment (pdunaj) [inline subsys/nfc/t4t/apdu.c]: if len is shorter shouldn't raw_data also be updated?

File: subsys/nfc/t4t/apdu.c [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+	/* Start to encode described CAPDU in the buffer. */
+	*raw_data++ = cmd_apdu->class_byte;
+	*raw_data++ = cmd_apdu->instruction;
+	*raw_data++ = MSB_16(cmd_apdu->parameter);
+	*raw_data++ = LSB_16(cmd_apdu->parameter);
+
+	/* Check if optional data field should be included. */
+	if (cmd_apdu->data.buff) {
+		/** Use long data length encoding. */
```
Comment (pdunaj) [inline subsys/nfc/t4t/apdu.c]: s/`/**` /`/*`

File: samples/nfc/tag_reader/src/main.c [hunk 10]
```diff
@@ -16,6 +16,7 @@
 #include <nfc/ndef/msg_parser.h>
```
Comment (b-gent) [inline samples/nfc/tag_reader/src/main.c]: Shouldn't we reflect this change in README.rst for this sample?

**Comment by KAGA164** [comment]
@pdunaj Can you marge ?

**Pull request merged**